src/main/java/org/elasticsearch/script/ScriptService.java

```diff
@@ -274,11 +278,30 @@ public CompiledScript compileInternal(Script script) {
         CompiledScript compiled = cache.getIfPresent(cacheKey);
         if (compiled == null) {
             //Either an un-cached inline script or an indexed script
-            compiled = new CompiledScript(lang, scriptEngineService.compile(code));
+            try {
+                compiled = new CompiledScript(script.getType(), name, lang, scriptEngineService.compile(code));
+            } catch (Exception e) {
+                String message = "Compilation error with " + script.getType() + " script ";
+                if (name != null) {
+                    message += "[" + name + "] ";
+                }
+                message += "using [" + lang + "]: " + ExceptionsHelper.detailedMessage(e);
+                throw new ScriptException(message, e);
+            }
             //Since the cache key is the script content itself we don't need to
             //invalidate/check the cache if an indexed script changes.
             cache.put(cacheKey, compiled);
         }
+
+        // An indexed or inline script may be mistaken for the other since the cache key relies on
+        // only the code of the script. Two indexed scripts may also have the same code, but different
+        // names. If this is case, the name and type parameters must be modified accordingly.
+        // Then cache the most recent version used.
```

Comment:
I guess the question is whether it's a problem that two scripts with same code end up being one single entry in the cache, beyond side effect of returning wrong name etc. (which is a bug). If so we should go for 2) and add all the needed info to the cache key so we can distinguish those "same" scripts. Otherwise we could go with 3) too. Both 2. and 3. sound ok to me, which one do you prefer?